node/network/collator-protocol/src/validator_side.rs

```diff
@@ -0,0 +1,1369 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, HashSet};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use futures::{
+	Future, SinkExt, StreamExt,
+	channel::{mpsc, oneshot}, stream::FuturesUnordered,
+};
+use futures_timer::Delay;
+use log::{trace, warn};
+
+use polkadot_primitives::v1::{
+	Id as ParaId, CandidateReceipt, CollatorId, Hash, PoV,
+};
+use polkadot_subsystem::{
+	FromOverseer, OverseerSignal, SubsystemContext,
+	messages::{
+		AllMessages, CollatorProtocolMessage, NetworkBridgeMessage,
+	},
+};
+use polkadot_node_network_protocol::{
+	v1 as protocol_v1, View, PeerId, ReputationChange as Rep, RequestId,
+	NetworkBridgeEvent,
+};
+
+use super::{modify_reputation, TARGET, Result};
+
+const COST_UNEXPECTED_MESSAGE: Rep = Rep::new(-10, "An unexpected message");
+const COST_REQUEST_TIMEDOUT: Rep = Rep::new(-20, "A collation request has timed out");
+const COST_REPORT_BAD: Rep = Rep::new(-50, "A collator was reported by another subsystem");
+const BENEFIT_NOTIFY_GOOD: Rep = Rep::new(50, "A collator was noted good by another subsystem");
+
+#[derive(Debug)]
+enum CollationRequestResult {
+	Received(RequestId),
+	Timeout(RequestId),
+}
+
+/// A Future representing an ongoing collation request.
+/// It may timeout or end in a graceful fashion if a requested
+/// collation has been received sucessfully or chain has moved on.
+struct CollationRequest {
+	// The response for this request has been received successfully or 
+	// chain has moved forward and this request is no longer relevant.
+	received: oneshot::Receiver<()>,
+
+	// The timeout of this request.
+	timeout: Delay,
+
+	// The id of this request.
+	request_id: RequestId,
+}
+
+impl Future for CollationRequest {
+	type Output = CollationRequestResult;
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll::<Self::Output> {
```

Comment:
Might be worth adding a comment to the effect that you can't just `select!` because this function is not async.

Reply:
Uhhhh in fact I will now try to remove this `Future`

Reply:
It might be possible to replace with an `async fn`